### PR TITLE
Show git-notes

### PR DESCRIPTION
**Pull request by CyberShadow**
This PR implements basic support for displaying notes created by [`git notes`](https://git-scm.com/docs/git-notes) (or software supporting it, such as magit).

The notes are read from Git's default ref location (`refs/notes/commits`), and thus it is assumed that the ref is pushed at that location to Gitea.

The implementation here is minimal; if the idea is OK, I can do any necessary changes.

### CODE REVIEW

**Comment by lunny** [comment]
Any screenshot? And could you change the text on template to `locale_en_US.ini`?

**Comment by zeripath** [comment]
I like the idea of this - however, is there any way to avoid reading the whole thing in memory and or making the page too big to render?

How about test the size of the note - if it's less than say 5Kb (most notes will be) then read it in otherwise provide a link that that the UI can asynchronously read. You can stick that as an API route and then you can test the note reading infrastructure relatively easy.

I think in general we should be more careful around ReadAll and page sizes.

**Comment by CyberShadow** [comment]
> Any screenshot?

Sure:


Although `git` displays notes directly beneath commit messages, I opted to place the notes outside the commit box in Gitea's UI, because their authorship is not tied to the commit's author/committer. Placing them right by the commit message would make it look like they were written by the same person as the rest of the commit.

> And could you change the text on template to `locale_en_US.ini`?

Done.

> I like the idea of this - however, is there any way to avoid reading the whole thing in memory and or making the page too big to render?

Sorry, could you please explain why you think this is necessary?

As far as I know, the typical use of git-notes is editable additions to commit messages. The `git-notes` manual states this as well, so, there's the implication that notes and commit messages shouldn't be treated too differently. I'm not familiar of a use case for git-notes which would involve placing large amounts of data in the notes objects.

Even though it's possible to create arbitrarily large git notes, it's also possible to create arbitrarily large commit messages. As far as I can see, Gitea will always display the entire commit message, without limiting it, and without providing a view to see it in its entirety. So, it doesn't make sense to me to do this for git notes without also doing it for commit messages, considering that there don't seem to be any pertinent distinctions between the two.

File: routers/repo/commit.go
Comment (lafriks) [inline routers/repo/commit.go]: In case there is error dataRc won't be closed

Comment (CyberShadow) [reply routers/repo/commit.go]: Fixed, thanks.

**Comment by lunny** [comment]
Could you display the author and time of the git note?

**Comment by CyberShadow** [comment]
I think it's possible, by looking at the last commit in the notes ref history which updated the note's path. 

Note that it would be more accurate to say that it would be the person who last edited the note, since if you want to add a note to a commit and a note already exists for that commit, you have no choice but to edit the existing note.

**Comment by CyberShadow** [comment]
> Could you display the author and time of the git note?

Done, how does this look?

**Comment by lafriks** [comment]
Could this be added in frame together with commit message?

**Comment by CyberShadow** [comment]
Something like this?



It doesn't look good in the dark theme, will likely need new CSS rules there.

**Comment by sapk** [comment]
AFAIK, notes are not signed by gpg/pgp so maybe it is better to display them under the signature.

**Comment by CyberShadow** [comment]
You certainly can sign the notes (by making signed commits in the notes ref), but that's a completely distinct signature from the commit that the notes are annotating.

**Review by sapk — APPROVED** [review]
This PR could be improved by :
- Filtering display of big note that could break ui
- Add tests to validate GetNote func
- Displaying that a note exist (like gpg lock) in commit list and a preview on passing over
- I personally would prefer to display notes like GPG. Attached to the commit  block to display that it is linked but under GPG (since itsn't verifed by it)

But this lay down some simple bases that works and could be improved later. And since it is not a main feature those changes are not blocking from my point of view. LGTM

**Comment by CyberShadow** [comment]
> Filtering display of big note that could break ui

I think that if we do this, it should also be done for commit messages.

> Add tests to validate GetNote func

Test added :)

> Displaying that a note exist (like gpg lock) in commit list and a preview on passing over

Good idea, I may submit a follow-up PR adding this.

> I personally would prefer to display notes like GPG. Attached to the commit block to display that it is linked but under GPG (since itsn't verifed by it)

Do you have a mockup of your idea? Would it be essentially as the second screenshot I posted but without the gap between the blocks?

**Comment by sapk** [comment]
> It should also be done for commit messages.

 Yes but that is for an other PR. It could be good to do it directly for notes but that isn't a blocking point (at least for me) as I suppose that people using note will understand and it could be fixed later. 

> Do you have a mockup of your idea? 

  I have done some testing here : https://github.com/sapk-fork/gitea/commit/2f453e724f8e4e8da462e27200e4bd94c3286fc9 
image
The main problem is that bottom class style does a lot to the GPG commit formatting. For testing I had to use inline css to counter that.

**Comment by lafriks** [comment]
Yes I was thinking about something like what @sapk did

**Comment by CyberShadow** [comment]
I like what you did with the header so I adapted that:

dark
light

I am not sure about gluing the note block to the commit block, for two reasons:

1. I think it looks a bit too clumped-up, it's not obvious to me at a first glance where the information about the commit ends and information about the note begins.
2. Hypothetically, a repository can contain multiple refs storing git notes. These can be due to having been written by different authors, or generated by different tools. It's possible that in the future users of Gitea might want to show notes from more than one ref (e.g. by having a list of refs as a per-repository config); in this case there probably should be one bubble per notes ref containing a note for the commit.

Of course you have the final word :)

**Comment by techknowlogick** [comment]
@CyberShadow I like what you did in the screenshots, as it will allow showing that notes are GPG signed similar to how commits are (to be done in a future PR of course).

**Comment by lafriks** [comment]
This looks really nice now

File: routers/repo/commit.go
Comment (lafriks) [inline routers/repo/commit.go]: This would be more correct:
```suggestion
	note := &git.Note{}
```

File: routers/repo/commit.go
Comment (lafriks) [inline routers/repo/commit.go]:
```suggestion
	err = git.GetNote(ctx.Repo.GitRepo, commitID, note)
```

**Comment by lafriks** [comment]
Those two small suggestions but otherwise looks good

File: routers/repo/commit.go
Comment (CyberShadow) [reply routers/repo/commit.go]: Thanks. I'm not familiar with Go and this is something that would not be allowed in other languages (taking the address of an rvalue). (Edit: found explanation [here](https://stackoverflow.com/a/40836095/21501))

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
Thanks for PR ;)

**Pull request merged**
